Python gettext localize (draft)

c-holtermann opened the pull request:
Localize (and parameterize) messages for the python console.

As suggested by John Ralls in https://github.com/Gnucash/gnucash/pull/501.
Bug report https://bugs.gnucash.org/show_bug.cgi?id=797228.

This is a first approach in this direction for the python parts. I'd like to get it right to continue for other python parts. I don't have much experience with gettext. Suggestions appreciated.

This is a draft for discussion. I'd like to keep it open for a bit.

c-holtermann commented on gnucash/python/init.py:
```diff
@@ -6,6 +6,8 @@
 from gi.repository import Gtk
 import os
+import gettext
+gettext.install("gnucash-pycons")
```
I just put in a random module name here. Is there a naming convention ?

c-holtermann replied:
It's gnucash. Changed that.

c-holtermann replied:
@fellen thanks for the hint about GETTEXT_PACKAGE. I suppose that that is not accessible through SWIG at the moment. I came upon the question of how much of the gnucash core data should be made accessible to python. I found the existing bindings sw_core_utils are a middleground to make selected functions accessible. I imagine the initialization of python gettext could also take place in a swig file where it is easier to access a range of internal stuff but keep it out of the python user sphere to provide a cleaner API.

c-holtermann replied:
I suppose it is also necessary to respect the possible locale changes in the gnucash environment files. I need to have a closer look at gnucash-bin.c. Maybe I can gnucash do the init/pref/environment stuff and copy the results like gettext.textdomain() etc. Or we just do the same things in python as we do in c.

In addition things work a bit differently in the python bindings than in the python shell module (I couldn't access environment variables in the latter even though gnc_environment_setup() gets called in gnucash_core.i and should have set them).

c-holtermann commented:
I haven't really worked with po files yet. Suggestions appreciated.

c-holtermann commented on gnucash/python/init.py:
```diff
@@ -6,6 +6,10 @@
+import gettext
+# install gettext _-function with path to mo files
+gettext.install("gnucash",
+        os.path.join(os.environ["GNC_HOME"],"share/locale"))
```
GNC_HOME is not available for the python bindings. So this will work only for the python shell. In addition I am not sure if GNC_HOME/share/locale will always be the right place.

c-holtermann replied:
There is a function gnc_path_get_localedir. I changed _sw_core_utils to make this accessible.

c-holtermann replied:
@fellen I tried to follow the way c and scheme use the gettext API. I didn't go the whole way though. As far as I understand the translation in python takes place in runtime (or at least possibly) so that the mo files in the target system are needed.

I hope that the question of the right path has been solved by https://github.com/Gnucash/gnucash/pull/506/commits/859808ecd7df4cb0a72a179daa9a4e2634f4a2dd.

I don't think that I overwrite something as the python gettext package just doesn't translate without setting a path. It does the same thing as c gettext but in a seperate way and needs a separate initialization. I don't think that it's possible in a simple way to access the exact same way c and scheme do by swig wrapping  those. But I may be wrong.

jralls replied:
Python's gettext module is just a wrapper around the gettext library.  Try this in pycons:
```python
>>> from gettext import textdomain
>>> textdomain()
```
If it says 'gnucash' there's no need to set it for pycons. Scripts using the bindings will of course have to set up gettext on their own.

c-holtermann replied:
textdomain() outputs "messages", not "gnucash". In pycons and in the python bindings.

fellen replied:
I have started https://wiki.gnucash.org/wiki/I18N#Using_Other_Programming_Languages for this project. See the Note there and https://github.com/Gnucash/gnucash/blob/maint/gnucash/gnucash-bin.c starting from the end.

jralls replied:
Looks like I'm wrong about python's [gettext module](https://github.com/python/cpython/blob/3.7/Lib/gettext.py). It's a reimplementation written mostly in python, so there's no way for it to share textdomain setting with the C library.

c-holtermann commented:
This is a working for my local installation (~/.local).

jralls commented on po/de.po:
```diff
@@ -20913,6 +20913,11 @@ msgstr "Konvertieren"
+#, python-brace-format
+msgid "Welcome to {0}!\n"
+msgstr "Willkommen zur {0}!\n"
+
```
I wonder if we should use "old style" python formatting, e.g.
```
(_('Welcome to %s') % 'GnuCash')
```
to increase the odds that we can re-use C msgids in Python.

c-holtermann replied:
> You can do it now for test purposes manually, but in the final version I would prefer, to have 3 separate commits:
> 
>     1. work on python
> 
>     2. merge commits up to xxx into de.po
> 
>     3. update de.po (with msgfmt's statistics as line 3)
I'll do that. I put the change in manually to have some data to test if it is working.

c-holtermann replied:
@jralls If using old-style formatting leads to a possibly bigger resource pool as translations have been made an can be reused I think that is a good idea. Maybe prefer old style and if not possible (because of complex formatting stuff) use newstyle as fallback.

jralls replied:
That sounds reasonable.

jralls commented:
The Travis failure isn't from this PR, it's a typo on the last maint commit, which I've fixed (I hope). @c-holtermann please rebase and force-push.

c-holtermann commented on gnucash/python/init.py:
```diff
@@ -98,9 +105,9 @@ def quit (self):
 if False:
     shelltype = "python"
-    title = "gnucash "+shelltype+" shell"
+    title = _("gnucash {0} shell").format(shelltype)
```
sure

c-holtermann commented on gnucash/python/init.py:
```diff
@@ -1,15 +1,16 @@
 import sys
 import gnucash._sw_app_utils as _sw_app_utils
+import gnucash._sw_core_utils as _sw_core_utils
```
I think I'd like to make sw_core_utils accessible in the python-bindings as well. That way this path to the locale files could be used there as well.

c-holtermann commented:
> 
> 
> The Travis failure isn't from this PR, it's a typo on the last maint commit, which I've fixed (I hope). @c-holtermann please rebase and force-push.

I thought so. Force pushed. Works.

c-holtermann commented:
It seems to me that I have to replicate the initialization done in gnucash-bin on the python side. I don't see how I can access that (see https://github.com/Gnucash/gnucash/pull/506#discussion_r281820316).

fellen commented on libgnucash/core-utils/core-utils.i:
```diff
@@ -59,6 +59,9 @@ gchar * gnc_path_get_bindir(void);
 %newobject gnc_path_get_localedir;
 gchar * gnc_path_get_localedir(void);
 
+/* Name of our gettext-domain (defined in config.h) */
```
JFYI: it is set in https://github.com/Gnucash/gnucash/blob/maint/CMakeLists.txt#L32 and then published via config.h

c-holtermann replied:
I saw that but I think that's specific enough for this comment.

fellen commented on gnucash/python/init.py:
```diff
@@ -9,7 +9,7 @@
 # install gettext for _-function, needs localepath
-gettext.install("gnucash",
+gettext.install(_sw_core_utils.GETTEXT_PACKAGE,
         _sw_core_utils.gnc_path_get_localedir())
```
I wonder if you can/want define shorthands for _sw_core_utils, but that has low priority.

fellen commented:
> It seems to me that I have to replicate the initialization done in gnucash-bin on the python side. I don't see how I can access that (see [#506 (comment)](https://github.com/Gnucash/gnucash/pull/506#discussion_r281820316)).

As this is anyhow working you could add a ToDo comment to keep it for later and continue with the other parts of the puzzle.

fellen commented on gnucash/python/init.py:
```diff
@@ -99,7 +99,7 @@ def refresh (self):
         """ quit """
 
-        self.write("\nHave a nice day !\n")
+        self.write(_("\nHave a nice day !\n"))
```
AFAiK only in French spaces before punctuation marks are in some cases common, in English not.

fellen replied:
... and in this specific I would prefer to reuse
28877:msgid "Have a nice day!"
and move the line feeds out of the translatable string.

c-holtermann commented on gnucash/python/init.py:
```diff
@@ -113,6 +113,7 @@ def quit (self):
     else:
         shelltypeName = "IPython"
     banner_style = 'title'
+    # TRANSLATORS: %s is either Python or IPython
```
As requested by @fellen.

c-holtermann commented:
I think the main issues have been taken care of. Thank you very much for your assistance. I will split this into different commits, moving the change to po file to a second pull request so that this one 1) implements the gettext scenario and 2) introduces some initial translatable strings. I don't yet understand how to do a proper .po commit but I will figure that out. How do I get rid of the "noise" mentionend in the wiki ?

c-holtermann commented:
There is an IF statement about gettext. What happens in the case that there is no gettext ? Should the _-function then just output the string ?

fellen commented:
> I think the main issues have been taken care of. Thank you very much for your assistance. I will split this into different commits, moving the change to po file to a second pull request so that this one 1) implements the gettext scenario and 2) introduces some initial translatable strings.

Right, but I would prefer to test the POT generation now, too.
> I don't yet understand how to do a proper .po commit but I will figure that out.

You should follow https://wiki.gnucash.org/wiki/Translation#Get_a_fresh_template. Probably we have to improve it or make pot.
> How do I get rid of the "noise" mentionend in the wiki ?

Separate it in it's own commit:
```
[git rebase]
make pot
msgmerge  ...
git commit ... "Merge recent changes into de.po ..."
# This will contain the noise

edit de.po
msgfmt -c ...
If errorfree git commit  ... "Update de.po ..."
# This will contain the real work

```

c-holtermann commented:
> 
> 
> > I think the main issues have been taken care of. Thank you very much for your assistance. I will split this into different commits, moving the change to po file to a second pull request so that this one 1) implements the gettext scenario and 2) introduces some initial translatable strings.
> 
> Right, but I would prefer to test the POT generation now, too.

I have added respective commits. Ok ?

> > I don't yet understand how to do a proper .po commit but I will figure that out.
> 
> You should follow https://wiki.gnucash.org/wiki/Translation#Get_a_fresh_template. Probably we have to improve it or make pot.

I have changed the wiki to reflect my understanding. Is that ok: https://wiki.gnucash.org/wiki/index.php?title=Translation&type=revision&diff=15355&oldid=15345 ?

c-holtermann commented:
> 
> 
> > There is an IF statement about gettext. What happens in the case that there is no gettext ? Should the _-function then just output the string ?
> 
> Yes

Ok, then I need to implement that check in python as well.

c-holtermann commented:
Ninja fails. I don't use it locally. So for the moment I ignore the problem.

fellen commented on po/POTFILES.in:
```diff
@@ -1,5 +1,26 @@
 # This is a list of files which contain translatable strings.
 # This file was autogenerated by cmake.
```
See this line! We need a pattern CMakeList.txt, which searchs for this files.

c-holtermann replied:
I'll remove this file. I'll change the line in CMakeList.txt. But the latter is just cosmetic as I have already included *.py.

fellen commented on libgnucash/core-utils/core-utils.i:
```diff
@@ -56,6 +56,15 @@ void
+
+/* Gettext availability */
+%constant int HAVE_GETTEXT = HAVE_GETTEXT;
+
```
I see no benefit here. It is set together with its relatives in https://github.com/Gnucash/gnucash/blob/c38c4557a56370a51236903e326f4429fd5b2caf/CMakeLists.txt#L700 ff
But if no (recent) version is found line 259 or 263 will abort CMake.

@gjanssens am I missing something her? Or should we remove all occurences of HAVE_GETTEXT like constants. It seems to me like a workaround, when certain feature were not available in some old versions of gettext or intltool.

fellen replied:
Another aspect: That check was intended for the C. I have several questions about Python:

If we talk about Python, https://en.wikipedia.org/wiki/CPython is meant on all platforms?

As it is part of their standard libs we need no tests in CMake for existence and/or compatibility?

jralls replied:
It's intended to support [Iron Python](https://ironpython.net/) too, but that also has gettext as a standard module. CMake tests won't do any good because python's gettext is needed at runtime, not build-time.

c-holtermann replied:
@fellen I have to redefine the constant to be able to access it from Python. I could also include config.h and access it directly but it seemed easier this way. If the flag is not necessary for python then that's a different issue. In that case this could be omitted as well.

fellen replied:
@c-holtermann As far as I understand it now, it is part of cPython's standard libs as it is in C's glib since a very long time. That has not been the case, when GnuCash has been a guile program "decades ago".

i am considering to remove HAVE_BIND_TEXTDOMAIN_CODESET, HAVE_DCGETTEXT and HAVE_GETTEXT from CMakeLists.txt, config.h[.in] and gnucash-bin.c, which is the only user, as it is only confusing today.
@jralls , @gjanssens any objections?

jralls replied:
@fellen go ahead. Only HAVE_GETTEXT is even used and it's effectively always defined: ```if (UNIX or MINGW)``` is all we support at present. If you're so motivated just clean out all of those settings.

fellen replied:
This Gettext flags were removed in commit 7cb4c1d.

c-holtermann replied:
@fellen no worries. I'm glad if work on one part leads to work on the bigger picture. Thanks. I'll adapt my files.

fellen commented:
From my POV you have now everything, which is required. You need only create a clean sequence of commits.

c-holtermann commented:
@fellen thanks. I think so too and will prepare a clean pull request. I want to to include a german translation of all translatable strings in this pull request as well.

jralls commented:
@c-holtermann Ping!  <3 weeks left until 3.6 release!

c-holtermann commented:
@jralls thanks for the reminder. I'm not sure if I will make it to 3.6. (even if it's not much to do) because I just moved to another town. Maybe next week. It would sure be good to finish this for the release.

cclauss commented:
Conflicting files po/de.po

c-holtermann commented:
I'm quite busy these days. I hope to get this finished in the coming weeks. Thanks for the hint.

gjanssens commented:
@c-holtermann Any chance you can do the final cleanup ? It'd be great to get this work merged in.

c-holtermann commented:
@gjanssens than you for asking and sorry for the delay. I will try to do find the time in the coming weeks. Life was full of surprising events and work.

c-holtermann commented:
I'm picking this up. I'll create a new PR with squashed commits of this PR. (python-gettext-localize-clean). I'll remove the changes to de.po and create a 2nd PR with translations to german. (python-gettext-localize-de). Is that ok?

c-holtermann commented:
Work continues on other PR mentioned above.
